crates/sui-core/src/authority_active.rs

```diff
@@ -176,33 +181,185 @@ impl<A> ActiveAuthority<A>
+
+    The active authority runs in a separate task, and manages the tasks of separate active
+    processes, with one sub-task per active process. These subtasks internally listen to a
+    watch channel on which others may send a start, shutdown, or update signal. When this
+    is received the tasks are restarted, or shut down.
+
+*/
+
+use tokio::sync::watch;
+
+pub enum LifecycleSignal {
+    Start,
+    Restart,
+    Exit,
+}
+
+pub struct LifecycleSignalSender {
+    sender: watch::Sender<LifecycleSignal>,
+}
+
+impl LifecycleSignalSender {
+    pub fn new() -> (Self, LifecycleTaskHandler) {
+        let (sender, receiver) = watch::channel(LifecycleSignal::Start);
+        (
+            LifecycleSignalSender { sender },
+            LifecycleTaskHandler { receiver },
+        )
+    }
+
+    pub async fn signal(&self, signal: LifecycleSignal) {
+        // TODO: reflect on what to do with this error.
+        let _ = self.sender.send(signal);
+    }
+}
+
+#[derive(Clone)]
+pub struct LifecycleTaskHandler {
+    pub receiver: watch::Receiver<LifecycleSignal>,
+}
+
+impl LifecycleTaskHandler {
+    pub async fn spawn<F, T>(mut self, description: String, fun: F)
+    where
+        F: Fn() -> T,
+        T: Future + Send + 'static,
+        <T as Future>::Output: Borrow<Result<(), SuiError>> + Send + 'static,
+    {
+        let mut handle: Option<tokio::task::JoinHandle<Result<(), SuiError>>> = None;
+        loop {
+            // flags for each iteration on whether we should
+            // start, restart or exit the task.
+            let mut start = false;
+            let mut abort = false;
+            let mut exit = false;
+
+            let signal = self.receiver.changed().await;
+            match signal {
+                Err(_err) => {
+                    info!("Closing active tasks, command channel was dropped.");
+                    abort = true;
+                    exit = true;
+                }
+                Ok(()) => {
+                    // We have actually received a new signal.
+                    match *self.receiver.borrow() {
+                        LifecycleSignal::Start => {
+                            info!("Starting task: {}", description);
+                            start = true;
+                        }
+                        LifecycleSignal::Restart => {
+                            info!("Restarting task: {}", description);
+                            abort = true;
+                            start = true;
+                        }
+                        LifecycleSignal::Exit => {
+                            info!("Exit task: {}", description);
+                            abort = true;
+                            exit = true;
+                        }
+                    }
+                }
+            }
+
+            // First close / abort previous task.
+            if abort {
+                if let Some(join_handle) = handle {
```

Comment:
What do you do with the handle? Is there any monitoring of if the task completes early, what should happen if this occurs, with or without and error?

Reply:
I have now added a `tokio::select` that checks whether the task if finished, and returns the result (which is a `Result<(), SuiError>`) for the moment.